### ISSUE TITLE
Arity of interface does not equal arity of implementation type

### ISSUE THREAD

**Issue by nathan5580** [dotnet/runtime#61407]
### Describe the bug

In .Net 5 I was able to have an interface implemented without the same arity.
But in .Net 6 it is no longer possible.

### To Reproduce

Interface
```
public interface IRepository<TEntity>
        where TEntity : class, new()
    {
    }
```

Implementation
```
public class Repository<TDbContext, TEntity> : IRepository<TEntity>
        where TEntity : class, new()
        where TDbContext : DbContext
    {
        protected readonly IDbContextFactory<TDbContext> Factory;

public Repository(IDbContextFactory<TDbContext> factory)
        {
            Factory = factory;
        }
    }
```

Then, inject this service
```
services.AddTransient(typeof(IRepository<>), typeof(Repository<,>))
```

### Exceptions (if any)

Exception:
```Arity of open generic service type 'Domain.Repositories.IRepository`1[TEntity]' does not equal arity of open generic implementation type 'Domain.Repositories.Repository`2[TDbContext,TEntity]'. (Parameter 'descriptors')```

Stacktrace:
   ```
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.Populate()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory..ctor(ICollection`1 descriptors)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(ICollection`1 serviceDescriptors, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.DefaultServiceProviderFactory.CreateServiceProvider(IServiceCollection containerBuilder)
   at Microsoft.Extensions.Hosting.Internal.ServiceFactoryAdapter`1.CreateServiceProvider(Object containerBuilder)
   at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at News.API.Program.Main(String[] args) in C:\Users\XXXXXXXXX\ISPB-BackEnd\News\News.API\Program.cs:line 32
```

### Further technical details
- ASP.NET Core version:
.Net 6
- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version:
Jetbrains rider (V2021.2.2) and VS Entreprise 2022 (17.1.0 Preview 1.0)
- Include the output of `dotnet --info`:

<details>
  <summary>dotnet --info Output</summary>

```
SDK .NET (reflétant tous les fichiers global.json) :
 Version:   6.0.100
 Commit:    9e8b04bbff

Environnement d'exécution :
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100\

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa

.NET SDKs installed:
  5.0.101 [C:\Program Files\dotnet\sdk]
  5.0.202 [C:\Program Files\dotnet\sdk]
  5.0.300 [C:\Program Files\dotnet\sdk]
  5.0.402 [C:\Program Files\dotnet\sdk]
  6.0.100-preview.5.21302.13 [C:\Program Files\dotnet\sdk]
  6.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

</details>

**Comment by KalleOlaviNiemitalo** [comment]
The arity check was added in <https://github.com/dotnet/runtime/pull/51167>. In .NET 5, if you registered an implementation with a different arity, was the dependency container able to instantiate it? <https://github.com/dotnet/runtime/issues/40848> claims it wasn't.

**Comment by nathan5580** [comment]
I do not really understand what to do to solve the problem, what is sure is that it worked before the migration of my project in .Net 6

**Comment by davidfowl** [comment]
AFAIK this never worked previously, it just failed when you were going to resolve the `IRepository<T>` instead of failing when you were adding registrations. 

```C#
using Microsoft.Extensions.DependencyInjection;

var services = new ServiceCollection();
services.AddSingleton(typeof(IRepository<>), typeof(Repository<,>));
var sp = services.BuildServiceProvider();

sp.GetRequiredService<IRepository<object>>();

public interface IRepository<TEntity>
        where TEntity : class, new()
{
}

public class Repository<TDbContext, TEntity> : IRepository<TEntity>
        where TEntity : class, new()
        where TDbContext : DbContext
{
    protected readonly IDbContextFactory<TDbContext> Factory;

    public Repository(IDbContextFactory<TDbContext> factory)
    {
        Factory = factory;
    }
}

public class DbContext
{

}

public interface IDbContextFactory<TDbContext>
{

}
```

This program fails when you call GetService on .NET 5 but fails when you call AddSingleton in .NET 6.

**Comment by nathan5580** [comment]
I confirm it is working on my .Net 5 environment.

I implement a new repository this way to use it

```
public interface ILogRepository : IRepository<Logs>
    {
    }

    public class LogRepository : Repository<CoreDbContext, Logs>, ILogRepository
    {
        public LogRepository(IDbContextFactory<CoreDbContext> factory) : base(factory)
        {
        }
    }
```

And register it like that

`services.AddTransient<ILogRepository, LogRepository>();
`

**Comment by nathan5580** [comment]
@davidfowl [Here ](https://github.com/nathan5580/GitHubArity)is the repo (https://github.com/nathan5580/GitHubArity)

Thank you

**Comment by davidfowl** [comment]
Thanks @nathan5580. It's unclear why you're adding the open generic registration when you're not even using it. That's likely why it's working in .NET 5. You're adding the open generic registration but you're never resolving an `IRepository<T>` in the application, so it's never hitting that code path. In fact, if you added an `IRepository<Type>` to any constructor, it would fail in in a similar way. This change should just catch errors earlier than you otherwise could before.

TL;DR you don't need to have that registration to make your scenario work.

**Comment by nathan5580** [comment]
I thought I needed it for this to work, I just tested taking it off and indeed all is working properly!
Thanks for your help @davidfowl

**Comment by markmillercc** [comment]
@davidfowl @nathan5580

Hello,

I have a scenario where this is causing an issue when using MediatR. I posted the issue there: https://github.com/jbogard/MediatR/issues/674

Simple reproduction here: https://github.com/markmillercc/ArityErrorReproduced

All the code is here: https://github.com/markmillercc/ArityErrorReproduced/blob/main/Test/Program.cs
and here is the test: https://github.com/markmillercc/ArityErrorReproduced/blob/main/Test.Tests/UnitTest1.cs

The unit test will fail if you upgrade Microsoft.Extensions.Hosting to 6.0.0.

I have a generic class and I am registering it closed with the concrete implementation. 

```c#
public class Program
{
    static void Main(string[] args)
    {
        Host.CreateDefaultBuilder(args)
            .ConfigureServices((hostContext, services) =>
            {
                services.AddMediatR(typeof(Program));
                services.AddTransient<INotificationHandler<MyEvent>, GenericEventHandler<MyEvent, MyCommand>>();
            })
            .Build()
            .Run();
    }
}

public class MyCommand { }

public class MyEvent : INotification { }

public class GenericEventHandler<TEvent, TCommand> : INotificationHandler<TEvent>
    where TEvent : INotification
{
    public Task Handle(TEvent notification, CancellationToken cancellationToken)
        => Task.CompletedTask;
}
```

However, behind the scenes in `AddMediatR()`, it is attempting to find that open generic and register that, which is now causing the error. 

I thought this would be a more common problem but maybe it's just me. Hoping to hear something back from the good folks at MediatR, but until then thought I would share it here.

If anyone has suggestions, I would love to hear them.

Thanks!
